### PR TITLE
Add devcontainer support to PyTorch Project

### PR DESCRIPTION
**Pull request by drisspg**
# Summary

### <samp>🤖 Generated by Copilot at 293ded1</samp>

This pull request adds support for using Visual Studio Code Remote - Containers extension with the pytorch project. It adds a `.devcontainer` folder with a `devcontainer.json` file, a `Dockerfile`, and a `noop.txt` file that configure and create a dev container with Anaconda and Python 3.

### <samp>🤖 Generated by Copilot at d6b9cd7</samp>

> _`devcontainer.json`_
> _Configures PyTorch containers_
> _For CPU or GPU_

## Related to:
https://github.com/pytorch/pytorch/issues/92838

cc @ZainRizvi @kit1980 @huydhn @clee2000

### CODE REVIEW

**Comment by drisspg** [comment]
Currently trying to figure out how opinionated I should be in the container. I prefer clang and lldb, hence some scripts to do that.  Ideally there would be some user defined mix-ins that can be added to customize to the users preferences, still trying to figure that.  @ZainRizvi  I would be curious for the devx take on this specifically around choice of base docker image, perhaps I should be using an existing CI reference? 

I have been running this on my local devcontainer and can build Pytorch with these settings:
```
export USE_MPS=1=0
export USE_KINETO=0
export BUILD_TEST=0
export USE_DISTRIBUTED=0
export BUILD_CAFFE2=0
export BUILD_CAFFE2_OPS=0
export USE_FBGEMM=0
export USE_OPENCV=0
export USE_QNNPACK=0
export USE_XNNPACK=0
export DEBUG=1
export USE_CUDA=0
```
Also since the cluster has Docker installed, this does work pretty seamlessly there as well and can build with same settings. I am going to create a cuda build

**Review by ZainRizvi — COMMENTED** [review]
Thanks for taking this on, it's a great PR to see!

I see you're following the instructions from README.md to setup this image. I do like the idea of using our existing CI base images instead for an even more consistent debugging experience though.  @huydhn, would you be able to point him to a good image to use?

cc: @malfet

File: update_alternatives_clang.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
```
Comment (malfet) [inline update_alternatives_clang.sh:1]: This should go int `scripts` or even `.devcontainer/scripts`?

File: update_alternatives_clang.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+# update_alternatives_clang.sh
+# chmod u+x update_alternatives_clang.sh
```
Comment (malfet) [inline update_alternatives_clang.sh:3]: Why not check it it as executable?

File: update_alternatives_clang.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#
+
+update_alternatives() {
+    local version=${1}
+    local priority=${2}
+    local master=${3}
+    local slaves=${4}
```
Comment (malfet) [inline update_alternatives_clang.sh:10]: And so on..
```suggestion
    local leader=${3}
    local follower=${4}
```

**Review by malfet — CHANGES_REQUESTED** [review]
Thank you for working on this.

I think base image shouldn't matter(as long as it reasonably recent), nor choice of compiler (though I guess VSCode is better at parsing `clang` errors than CPP ones

Please move `update_laternatives` to `.devicontainer/scripts` folder rather than keep it in top level directory.

Also, I'm curious, why  not have single Dockerfile, with an extra step for CUDA (which is triggered via arg in devcontainer.json)

File: update_alternatives_clang.sh (hunk 3 above)
Comment (drisspg) [reply update_alternatives_clang.sh:10]: while I don't agree with the terminology, it is very present in the man pages for update-alternatives: https://man7.org/linux/man-pages/man1/update-alternatives.1.html

Comment (malfet) [reply update_alternatives_clang.sh:10]: Ok. But please move this script to `.devcontainer` folder. Or perhaps delete it altogether, if you don't plan to configure clang as default compiler.

File: .devcontainer/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# Can't build succesfully, The exports need to go in the .sh script
+# But also clang-15 might be two new for cuda 11.8
+
+# ARG USE_CLANG
+# RUN if [ -n "$USE_CLANG" ]; then \
+#     sudo bash -c "$(wget -O - https://apt.llvm.org/llvm.sh)"; \
+#     sudo .devcontainer/scripts/update_alternatives_clang.sh 15 10; \
+#     export CC=clang; \
+#     export CXX=clang++; \
+#     fi
```
Comment (ZainRizvi) [inline .devcontainer/Dockerfile:27]: Let's not merge commented out code

File: .devcontainer/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+# This installs llvm toolchain
+
+# Instal cuda if version is specified
+ARG CUDA_VERSION
+RUN if [ -n "$CUDA_VERSION" ]; then \
+       conda install cuda -c "nvidia/label/cuda-${CUDA_VERSION}"; \
+    fi
+# [Optional] Uncomment this section to install additional OS packages.
+# RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
+#     && apt-get -y install --no-install-recommends <your-package-list-here>
```
Comment (ZainRizvi) [inline .devcontainer/Dockerfile:38]: What about turning this into an ARG instead, like the other sections?

**Review by ZainRizvi — APPROVED** [review]

File: .devcontainer/Dockerfile (hunk 4 above)
Comment (drisspg) [reply .devcontainer/Dockerfile]: I have tested this workflow I can build with clang succesfully on CPU but when trying to compile with CUDA: 
```

  #$ "/usr/bin"/clang -D__CUDA_ARCH__=520 -D__CUDA_ARCH_LIST__=520 -E -x c++
  -DCUDA_DOUBLE_MATH_FUNCTIONS -D__CUDACC__ -D__NVCC__
  "-I/opt/conda/bin/../include" -D__CUDACC_VER_MAJOR__=11
  -D__CUDACC_VER_MINOR__=8 -D__CUDACC_VER_BUILD__=89
  -D__CUDA_API_VER_MAJOR__=11 -D__CUDA_API_VER_MINOR__=8
  -D__NVCC_DIAG_PRAGMA_SUPPORT__=1 -include "cuda_runtime.h" -m64
  "CMakeCUDACompilerId.cu" -o "tmp/CMakeCUDACompilerId.cpp1.ii"

  /usr/bin/clang: /opt/conda/bin/../lib/libtinfo.so.6: no version information
  available (required by /usr/lib/x86_64-linux-gnu/libLLVM-11.so.1)

  /usr/bin/clang: /opt/conda/bin/../lib/libtinfo.so.6: no version information
  available (required by /usr/lib/x86_64-linux-gnu/libedit.so.2)

  /usr/bin/clang: symbol lookup error:
  /usr/lib/x86_64-linux-gnu/libLLVM-11.so.1: undefined symbol:
  ffi_type_sint32, version LIBFFI_BASE_7.0

  # --error 0x7f --

  

  

Call Stack (most recent call first):
  /opt/conda/share/cmake-3.22/Modules/CMakeDetermineCompilerId.cmake:6 (CMAKE_DETERMINE_COMPILER_ID_BUILD)
  /opt/conda/share/cmake-3.22/Modules/CMakeDetermineCompilerId.cmake:48 (__determine_compiler_id_test)
  /opt/conda/share/cmake-3.22/Modules/CMakeDetermineCUDACompiler.cmake:298 (CMAKE_DETERMINE_COMPILER_ID)
  cmake/public/cuda.cmake:47 (enable_language)
  cmake/Dependencies.cmake:43 (include)
  CMakeLists.txt:710 (include)
  ```
  
  I think this has to do with the mismatch between conda install cuda libs.  I can remove this Clang stuff entirely but not sure

File: .devcontainer/Dockerfile (hunk 5 above)
Comment (drisspg) [reply .devcontainer/Dockerfile]: Good idea will add

**Comment by GZ82** [comment]
Thanks very much for add this! is it possible to set up one for MacOS plus AMD GPU?

**Comment by drisspg** [comment]
@GZ82 
Definitely, would you like to contribute this? I will create an issue tracking this

**Pull request closed without merging**